File: src/examples/python/hooking_libc.py
#!/usr/bin/env python2
## -*- coding: utf-8 -*-
##
## Output:
##
##  $ python ./src/examples/python/hooking_libc.py
##  [+] Loading 0x400040 - 0x400270
##  [+] Loading 0x400270 - 0x40028c
##  [+] Loading 0x400000 - 0x4007a4
##  [+] Loading 0x600e10 - 0x601048
##  [+] Loading 0x600e28 - 0x600ff8
##  [+] Loading 0x40028c - 0x4002ac
##  [+] Loading 0x400678 - 0x4006ac
##  [+] Loading 0x000000 - 0x000000
##  [+] Loading 0x600e10 - 0x601000
##  [+] Loading 0x000000 - 0x000000
##  [+] Hooking strlen
##  [+] Hooking printf
##  [+] Hooking __libc_start_main
##  [+] Starting emulation.
##  4004a0: xor ebp, ebp
##  4004a2: mov r9, rdx
##  4004a5: pop rsi
##  4004a6: mov rdx, rsp
##  4004a9: and rsp, 0xfffffffffffffff0
##  4004ad: push rax
##  4004ae: push rsp
##  4004af: mov r8, 0x400650
##  4004b6: mov rcx, 0x4005e0
##  4004bd: mov rdi, 0x400596
##  4004c4: call 0x400480
##  400480: jmp qword ptr [rip + 0x200ba2]
##  [+] __libc_start_main hooked
##  400596: push rbp
##  400597: mov rbp, rsp
##  40059a: sub rsp, 0x10
##  40059e: mov dword ptr [rbp - 4], edi
##  4005a1: mov qword ptr [rbp - 0x10], rsi
##  4005a5: cmp dword ptr [rbp - 4], 2
##  4005a9: je 0x4005b2
##  4005b2: mov rax, qword ptr [rbp - 0x10]
##  4005b6: add rax, 8
##  4005ba: mov rax, qword ptr [rax]
##  4005bd: mov rdi, rax
##  4005c0: call 0x400460
##  400460: jmp qword ptr [rip + 0x200bb2]
##  [+] Strlen hooked
##  4005c5: mov rsi, rax
##  4005c8: mov edi, 0x400664
##  4005cd: mov eax, 0
##  4005d2: call 0x400470
##  400470: jmp qword ptr [rip + 0x200baa]
##  [+] printf hooked
##  Input size = 12
##  4005d7: mov eax, 0
##  4005dc: leave
##  4005dd: ret
##  4004c9: hlt
##  [+] Emulation done
##

import os
import sys
import string

from triton import TritonContext, ARCH, CPUSIZE, MemoryAccess, OPCODE, Instruction

Triton = TritonContext()


def getMemoryString(addr):
    s = str()
    index = 0

    while Triton.getConcreteMemoryValue(addr+index):
        c = chr(Triton.getConcreteMemoryValue(addr+index))
        if c not in string.printable: c = ""
        s += c
        index  += 1

    return s


def getFormatString(addr):
    return getMemoryString(addr)                                                    \
           .replace("%s", "{}").replace("%d", "{:d}").replace("%#02x", "{:#02x}")   \
           .replace("%#x", "{:#x}").replace("%x", "{:x}").replace("%02X", "{:02x}") \
           .replace("%c", "{:c}").replace("%02x", "{:02x}").replace("%ld", "{:d}")  \
           .replace("%*s", "").replace("%lX", "{:x}").replace("%08x", "{:08x}")     \
           .replace("%u", "{:d}")                                                   \


# Simulate the strlen() function
def strlenHandler():
    print '[+] Strlen hooked'

    # Get arguments
    arg1 = getMemoryString(Triton.getConcreteRegisterValue(Triton.registers.rdi))

    # Return value
    return len(arg1)


# Simulate the printf() function
def printfHandler():
    print '[+] printf hooked'

    # Get arguments
    arg1   = getFormatString(Triton.getConcreteRegisterValue(Triton.registers.rdi))
    arg2   = Triton.getConcreteRegisterValue(Triton.registers.rsi)
    arg3   = Triton.getConcreteRegisterValue(Triton.registers.rdx)
    arg4   = Triton.getConcreteRegisterValue(Triton.registers.rcx)
    arg5   = Triton.getConcreteRegisterValue(Triton.registers.r8)
    arg6   = Triton.getConcreteRegisterValue(Triton.registers.r9)
    nbArgs = arg1.count("{")
    args   = [arg2, arg3, arg4, arg5, arg6][:nbArgs]
    s      = arg1.format(*args)

    sys.stdout.write(s)

    # Return value
    return len(s)


def libcMainHandler():
    print '[+] __libc_start_main hooked'

    # Get arguments
    main = Triton.getConcreteRegisterValue(Triton.registers.rdi)

    # Push the return value to jump into the main() function
    Triton.concretizeRegister(Triton.registers.rsp)
    Triton.setConcreteRegisterValue(Triton.registers.rsp, Triton.getConcreteRegisterValue(Triton.registers.rsp)-CPUSIZE.QWORD)

    ret2main = MemoryAccess(Triton.getConcreteRegisterValue(Triton.registers.rsp), CPUSIZE.QWORD)
    Triton.concretizeMemory(ret2main)
    Triton.setConcreteMemoryValue(ret2main, main)

    # Setup argc / argv
    Triton.concretizeRegister(Triton.registers.rdi)
    Triton.concretizeRegister(Triton.registers.rsi)

    argvs = [
        "sample_1",      # argv[0]
        "my_first_arg",  # argv[1]
    ]

    # Define argc / argv
    base  = 0x20000000
    addrs = list()

    for argv in argvs:
        addrs.append(base)
        Triton.setConcreteMemoryAreaValue(base, argv+'\x00')
        base += len(argv)+1

    argc = len(argvs)
    argv = base
    for addr in addrs:
        Triton.setConcreteMemoryValue(MemoryAccess(base, CPUSIZE.QWORD), addr)
        base += CPUSIZE.QWORD

    Triton.setConcreteRegisterValue(Triton.registers.rdi, argc)
    Triton.setConcreteRegisterValue(Triton.registers.rsi, argv)

    return 0


customRelocation = [
    ('strlen',            strlenHandler,   0x10000000),
    ('printf',            printfHandler,   0x10000001),
    ('__libc_start_main', libcMainHandler, 0x10000002),
]


def hookingHandler():
    pc = Triton.getConcreteRegisterValue(Triton.registers.rip)
    for rel in customRelocation:
        if rel[2] == pc:
            # Emulate the routine and the return value
            ret_value = rel[1]()
            Triton.concretizeRegister(Triton.registers.rax)
            Triton.setConcreteRegisterValue(Triton.registers.rax, ret_value)

            # Get the return address
            ret_addr = Triton.getConcreteMemoryValue(MemoryAccess(Triton.getConcreteRegisterValue(Triton.registers.rsp), CPUSIZE.QWORD))

            # Hijack RIP to skip the call
            Triton.concretizeRegister(Triton.registers.rip)
            Triton.setConcreteRegisterValue(Triton.registers.rip, ret_addr)

            # Restore RSP (simulate the ret)
            Triton.concretizeRegister(Triton.registers.rsp)
            Triton.setConcreteRegisterValue(Triton.registers.rsp, Triton.getConcreteRegisterValue(Triton.registers.rsp)+CPUSIZE.QWORD)
    return


# Emulate the CheckSolution() function.
def emulate(pc):
    print '[+] Starting emulation.'

    while pc:
        # Fetch opcode
        opcode = Triton.getConcreteMemoryAreaValue(pc, 16)

        # Create the Triton instruction
        instruction = Instruction()
        instruction.setOpcode(opcode)
        instruction.setAddress(pc)

        # Process
        Triton.processing(instruction)
        print instruction

        if instruction.getType() == OPCODE.HLT:
            break

        # Simulate routines
        hookingHandler()

        # Next
        pc = Triton.getConcreteRegisterValue(Triton.registers.rip)

    print '[+] Emulation done.'
    return


def loadBinary(path):
    import lief
    binary = lief.parse(path)
    phdrs  = binary.segments
    for phdr in phdrs:
        size   = phdr.physical_size
        vaddr  = phdr.virtual_address
        print '[+] Loading 0x%06x - 0x%06x' %(vaddr, vaddr+size)
        Triton.setConcreteMemoryAreaValue(vaddr, phdr.content)
    return binary


def makeRelocation(binary):
    # Perform our own relocations
    for rel in binary.pltgot_relocations:
        symbolName = rel.symbol.name
        symbolRelo = rel.address
        for crel in customRelocation:
            if symbolName == crel[0]:
                print '[+] Hooking %s' %(symbolName)
                Triton.setConcreteMemoryValue(MemoryAccess(symbolRelo, CPUSIZE.QWORD), crel[2])
    return


if __name__ == '__main__':
    # Set the architecture
    Triton.setArchitecture(ARCH.X86_64)

    # Load the binary
    binary = loadBinary(os.path.join(os.path.dirname(__file__), 'samples', 'sample_1'))

    # Perform our own relocations
    makeRelocation(binary)

    # Define a fake stack
    Triton.setConcreteRegisterValue(Triton.registers.rbp, 0x7fffffff)
    Triton.setConcreteRegisterValue(Triton.registers.rsp, 0x6fffffff)

    # Let's emulate the binary from the entry point
    emulate(binary.entrypoint)

    sys.exit(0)

File: src/examples/cpp/hooking_libc.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <functional>
#include <triton/context.hpp>
#include <triton/x86Specifications.hpp>
#include <LIEF/ELF.hpp>
#include <fmt/format.h>
#include <sstream>
#include <iomanip>

#define DEBUG 0

std::string getMemoryString(const triton::Context& ctx, triton::uint64 addr) {
    std::string s;
    size_t index = 0;

    while (auto value = ctx.getConcreteMemoryValue(addr + index)) {
        char c = static_cast<char>(value);
        if (std::isprint(c)) {
            s += c;
        }
        index++;
    }

    return s;
}

std::string getFormatString(const triton::Context& ctx, triton::uint64 addr) {
    std::string s = getMemoryString(ctx, addr);
    // Implement string replacements here if needed
    return s;
}

triton::uint64 strlenHandler(triton::Context& ctx) {
    if constexpr (DEBUG) {
        std::cout << "[+] Strlen hooked" << std::endl;
    }
    auto arg1 = getMemoryString(ctx, ctx.getConcreteRegisterValue(ctx.registers.x86_rdi).convert_to<triton::uint64>());
    return arg1.length();
}



triton::uint64 printfHandler(triton::Context& ctx) {
    if constexpr (DEBUG) {
        std::cout << "[+] printf hooked" << std::endl;
    }
    auto format_str = getFormatString(ctx, ctx.getConcreteRegisterValue(ctx.registers.x86_rdi).convert_to<triton::uint64>());
    auto arg1 = ctx.getConcreteRegisterValue(ctx.registers.x86_rsi);

    std::stringstream ss;
    size_t pos = format_str.find("%ld");
    if (pos != std::string::npos) {
        ss << format_str.substr(0, pos) << arg1 << format_str.substr(pos + 3);
    } else {
        ss << format_str;
    }

    std::string formatted_str = ss.str();
    std::cout << formatted_str << std::endl;
    return formatted_str.length();
}

triton::uint64 libcMainHandler(triton::Context& ctx) {
    if constexpr (DEBUG) {
        std::cout << "[+] __libc_start_main hooked" << std::endl;
    }

    auto main = ctx.getConcreteRegisterValue(ctx.registers.x86_rdi);

    ctx.setConcreteRegisterValue(ctx.registers.x86_rsp,
        ctx.getConcreteRegisterValue(ctx.registers.x86_rsp) - 8);

    auto ret2main = triton::arch::MemoryAccess(ctx.getConcreteRegisterValue(ctx.registers.x86_rsp).convert_to<triton::uint64>(), 8);
    ctx.setConcreteMemoryValue(ret2main, main);

    ctx.concretizeRegister(ctx.registers.x86_rdi);
    ctx.concretizeRegister(ctx.registers.x86_rsi);

    std::vector<std::string> argvs = {"sample_1", "my_first_arg"};

    triton::uint64 base = 0x20000000;
    std::vector<triton::uint64> addrs;

    for (const auto& argv : argvs) {
        addrs.push_back(base);
        ctx.setConcreteMemoryAreaValue(base, reinterpret_cast<const triton::uint8*>(argv.c_str()), argv.length() + 1);
        base += argv.length() + 1;
    }

    triton::uint64 argc = argvs.size();
    triton::uint64 argv = base;
    for (const auto& addr : addrs) {
        ctx.setConcreteMemoryValue(triton::arch::MemoryAccess(base, 8), addr);
        base += 8;
    }

    ctx.setConcreteRegisterValue(ctx.registers.x86_rdi, argc);
    ctx.setConcreteRegisterValue(ctx.registers.x86_rsi, argv);

    return 0;
}

struct CustomRelocation {
    std::string name;
    std::function<triton::uint64(triton::Context&)> handler;
    triton::uint64 address;
};

std::vector<CustomRelocation> customRelocation = {
    {"strlen", strlenHandler, 0x10000000},
    {"printf", printfHandler, 0x10000001},
    {"__libc_start_main", libcMainHandler, 0x10000002},
};

void hookingHandler(triton::Context& ctx) {
    auto pc = ctx.getConcreteRegisterValue(ctx.registers.x86_rip);
    for (const auto& rel : customRelocation) {
        if (rel.address == pc) {
            auto ret_value = rel.handler(ctx);
            ctx.setConcreteRegisterValue(ctx.registers.x86_rax, ret_value);

            auto ret_addr = ctx.getConcreteMemoryValue(triton::arch::MemoryAccess(
                ctx.getConcreteRegisterValue(ctx.registers.x86_rsp).convert_to<triton::uint64>(), 8));

            ctx.setConcreteRegisterValue(ctx.registers.x86_rip, ret_addr);
            ctx.setConcreteRegisterValue(ctx.registers.x86_rsp,
                ctx.getConcreteRegisterValue(ctx.registers.x86_rsp) + 8);
        }
    }
}

void emulate(triton::Context& ctx, triton::uint64 pc) {
    if constexpr (DEBUG) {
        std::cout << "[+] Starting emulation." << std::endl;
    }

    while (pc) {
        auto opcode = ctx.getConcreteMemoryAreaValue(pc, 16);

        triton::arch::Instruction instruction;
        instruction.setOpcode(opcode.data(), opcode.size());
        instruction.setAddress(pc);

        ctx.processing(instruction);
        if constexpr (DEBUG) {
            std::cout << instruction << std::endl;
        }

        if (instruction.getType() == triton::arch::x86::ID_INS_HLT || instruction.getType() == triton::arch::x86::ID_INS_RET) { // add ID_INS_RET too maybe? see issue https://github.com/JonathanSalwan/Triton/issues/912
            break;
        }

        hookingHandler(ctx);

        pc = ctx.getConcreteRegisterValue(ctx.registers.x86_rip).convert_to<triton::uint64>();
    }
    if constexpr (DEBUG) {
        std::cout << "[+] Emulation done." << std::endl;
    }
}

std::unique_ptr<LIEF::ELF::Binary> loadBinary(triton::Context& ctx, const std::string& path) {
    std::unique_ptr<LIEF::ELF::Binary> binary = LIEF::ELF::Parser::parse(path);
    if (!binary) {
        throw std::runtime_error("Failed to parse ELF binary");
    }

    for (const auto& segment : binary->segments()) {
        if (segment.type() != LIEF::ELF::Segment::TYPE::LOAD) {
            continue;
        }
        triton::uint64 vaddr = segment.virtual_address();
        const auto& content = segment.content(); // This is now a span<const uint8_t>
        if constexpr (DEBUG) {
            std::cout << "[+] Loading 0x" << std::hex << vaddr << " - 0x" << (vaddr + content.size()) << std::endl;
        }

        // Use data() and size() methods of span
        ctx.setConcreteMemoryAreaValue(vaddr, content.data(), content.size());
    }

    return binary;
}

void makeRelocation(triton::Context& ctx, const LIEF::ELF::Binary& binary) {
    for (const auto& pltgot_entry : binary.pltgot_relocations()) {
        const std::string& symbolName = pltgot_entry.symbol()->name();
        triton::uint64 symbolRelo = pltgot_entry.address();
        for (const auto& crel : customRelocation) {
            if (symbolName == crel.name) {
                if constexpr (DEBUG) {
                    std::cout << "[+] Hooking " << symbolName << std::endl;
                }
                ctx.setConcreteMemoryValue(triton::arch::MemoryAccess(symbolRelo, 8), crel.address);
            }
        }
    }
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <path_to_binary>" << std::endl;
        return 1;
    }

    triton::Context ctx;
    ctx.setArchitecture(triton::arch::ARCH_X86_64);

    std::unique_ptr<LIEF::ELF::Binary> binary;
    try {
        binary = loadBinary(ctx, argv[1]);
    } catch (const std::exception& e) {
        std::cerr << "Error loading binary: " << e.what() << std::endl;
        return 1;
    }

    makeRelocation(ctx, *binary);

    ctx.setConcreteRegisterValue(ctx.registers.x86_rbp, 0x7fffffff);
    ctx.setConcreteRegisterValue(ctx.registers.x86_rsp, 0x6fffffff);

    triton::uint64 entrypoint = binary->entrypoint();
    emulate(ctx, entrypoint);

    return 0;
}